### ISSUE TITLE
IKE Evaluation

### ISSUE THREAD

**Issue by shariqahn** [zjunlp/EasyEdit#459]
I would like to clarify my understanding of how you evaluate IKE on the test set. Looking at examples/run_zsre_llama2.py for example, it seems like you only use the saved context and embeddings from the training set when performing edits on the test set. Then, each prompt in the test set is added to icl_examples. Is that correct that you don't use the IKE template (copy, update, retain) to add the test edits to the context? Do you add to icl_examples continuously with each test edit or create a new array each time?

I am wondering because I found that performance of IKE was similar to the non-edited model when performing my own evaluations, so I would like to know the best way to evaluate performance of the edited IKE model despite the fact that it does not change model parameters directly.

**Comment by littlefive5** [comment]
Yes, we do not use the IKE template (copy, update, retain) to add the test edits to the context.
We create a new array each time for each edit case.

Here, `Similar' means it do not successfully edit the knowledge?

**Comment by zxlzr** [comment]
hi, could you please add more details of your questions, do you have any further questions?

**Comment by shariqahn** [comment]
The metric summary looks good, so I think the edits are successful:
```
Metrics Summary:  {'pre': {'rewrite_acc': 0.0, 'rephrase_acc': 0.075, 'locality': {'neighborhood_acc': 1.0}}, 'post': {'rewrite_acc': 1.0, 'rephrase_acc': 1.0, 'locality': {'neighborhood_acc': 0.8881953943785973}}}
```
I think the issue is more likely that I am saving or evaluating the model wrong because the edited model didn't change much from the baseline in my metrics for my experiments. Since the model parameters aren't changed in IKE, I think it makes sense that simply running my metrics in the saved model from:
`metrics, edited_model, _ = editor.edit(...)` and `edited_model.save_pretrained(model_save_dir)` would not be a good way to evaluate results. 

However, I am unsure what a better alternative would be. Is what you're saying that I should add the k nearest neighbors in the `icl_examples` array (which is all the IKE templates from all the training set edits) to each model generation on the test set? Like you did in https://github.com/zjunlp/EasyEdit/blob/main/easyeditor/models/ike/ike_main.py#L33C2-L53C88? But it seems wrong to me to apply the same logic performed during editing for evaluation. I would think evaluation should not involve further editing logic.

**Comment by littlefive5** [comment]
Hello, currently in our repo we would just add the k nearest neighbors in the `icl_examples` array (which is all the IKE templates from all the training set edits) to each model generation on the evaluation dataset.
I think it makes sense when we test other cases for evaluation we should not add edit examples as input. 
Maybe you can add a classifier or sematic filtering to determine whether to retrieve the examples for the coming input. It is a common way like SERAC did.

**Comment by shariqahn** [comment]
Just to make sure I understand, are you saying you add the neighbors from the training set edits during evaluation, but you do not recommend this approach? If it is true that it is not best practice to use this approach, I don't quite understand what the classifier or sematic filtering would be checking for since by that logic, there is never a situation where the neighbors should be added to the input. Unless you are saying that this approach is okay for inputs within the same distribution as the training set but not for inputs out-of-scope from the edits like SERAC does?

**Comment by littlefive5** [comment]
> Unless you are saying that appending the training set neighbors is okay for edited inputs but not for inputs out-of-scope from the edits like SERAC does?

yes, it is what I mean.
It's just my personal opinion that appending the training set neighbors is okay for edited inputs but not for inputs out-of-scope.
I think for retrieval-augumented editing, you do not need to append these examples for all the inputs.
However, currently, retrieval or classifier may not work that well. For example, you may need this knowledge during some middle reasoning steps, but sometimes, the classifier may not recognize this and say the input would not need to reach edit memory.

It's an open question here.

**Comment by zxlzr** [comment]
hi, do you have any further questions?

**Comment by shariqahn** [comment]
I am still working on refactoring my code to test this. I'll get back to you!
